File: macos/Tests/VoiceKeyCoreTests/DictationModeTests.swift
import Testing
@testable import VoiceKeyCore

@Suite struct DictationModeTests {

    // MARK: - Dictating while VoiceKey itself is frontmost

    @Test func aDictationIsInsertedIntoWhateverOtherAppIsFrontmost() {
        #expect(DictationMode.insertsAtCursor(frontmostBundleID: "com.apple.TextEdit",
                                              ownBundleID: "com.timherglotz.voicekey"))
    }

    @Test func aDictationIsNotInsertedWhenVoiceKeyItselfIsFrontmost() {
        // Otherwise it lands in the window's own search field, which filters the
        // library down to the transcript just spoken — it reads as data loss.
        #expect(!DictationMode.insertsAtCursor(frontmostBundleID: "com.timherglotz.voicekey",
                                               ownBundleID: "com.timherglotz.voicekey"))
    }

    @Test func theFrontmostAppIsRecognisedWhateverCaseItReportsItsBundleIDIn() {
        #expect(!DictationMode.insertsAtCursor(frontmostBundleID: "com.TimHerglotz.VoiceKey",
                                               ownBundleID: "com.timherglotz.voicekey"))
    }

    @Test func anUnknownFrontmostAppIsStillInsertedInto() {
        // A missing bundle ID is some other process, not us.
        #expect(DictationMode.insertsAtCursor(frontmostBundleID: nil,
                                              ownBundleID: "com.timherglotz.voicekey"))
    }

    @Test func anUnknownOwnBundleIDNeverSuppressesTheInsertion() {
        #expect(DictationMode.insertsAtCursor(frontmostBundleID: "com.timherglotz.voicekey",
                                              ownBundleID: nil))
    }

    // MARK: - Bundle ID → mode mapping

    @Test func terminalsAndEditorsMapToCodeMode() {
        let codeBundleIDs = [
            "com.apple.Terminal",
            "com.googlecode.iterm2",
            "com.mitchellh.ghostty",
            "com.microsoft.VSCode",
            "com.todesktop.230313mzl4w4u92",
            "com.apple.dt.Xcode",
            "dev.zed.Zed",
            "com.sublimetext.4",
            "com.github.GitHubClient",
        ]

        for bundleID in codeBundleIDs {
            #expect(DictationMode.mode(forBundleID: bundleID) == .code, "expected .code for \(bundleID)")
        }
    }

    @Test func jetBrainsIDEsMapToCodeModeByPrefix() {
        #expect(DictationMode.mode(forBundleID: "com.jetbrains.intellij") == .code)
    }

    @Test func mailClientsMapToEmailMode() {
        let emailBundleIDs = [
            "com.apple.mail",
            "com.microsoft.Outlook",
            "com.readdle.SparkDesktop",
            "com.superhuman.electron",
        ]

        for bundleID in emailBundleIDs {
            #expect(DictationMode.mode(forBundleID: bundleID) == .email, "expected .email for \(bundleID)")
        }
    }

    @Test func unknownAppsMapToStandardMode() {
        #expect(DictationMode.mode(forBundleID: "com.apple.Safari") == .standard)
        #expect(DictationMode.mode(forBundleID: "com.tinyspeck.slackmacgap") == .standard)
    }

    @Test func missingBundleIDMapsToStandardMode() {
        #expect(DictationMode.mode(forBundleID: nil) == .standard)
    }

    @Test func bundleIDMatchingIgnoresCase() {
        #expect(DictationMode.mode(forBundleID: "COM.APPLE.TERMINAL") == .code)
        #expect(DictationMode.mode(forBundleID: "com.apple.terminal") == .code)
        #expect(DictationMode.mode(forBundleID: "COM.JetBrains.IntelliJ") == .code)
        #expect(DictationMode.mode(forBundleID: "COM.APPLE.MAIL") == .email)
    }

    @Test func jetBrainsPrefixDoesNotMatchLookalikeBundleIDs() {
        #expect(DictationMode.mode(forBundleID: "com.jetbrainsfan.notanide") == .standard)
    }

    // MARK: - Code mode formatting

    @Test func codeModeStripsSingleTrailingPeriod() {
        #expect(DictationMode.code.format("git status.") == "git status")
    }

    @Test func codeModePreservesEllipsisAndRepeatedDots() {
        #expect(DictationMode.code.format("wait...") == "wait...")
        #expect(DictationMode.code.format("wait…") == "wait…")
    }

    @Test func codeModePreservesOtherTerminalPunctuation() {
        #expect(DictationMode.code.format("really?") == "really?")
        #expect(DictationMode.code.format("stop!") == "stop!")
    }

    @Test func codeModePreservesInteriorPeriodsAndCasing() {
        #expect(DictationMode.code.format("node.js is fine.") == "node.js is fine")
        #expect(DictationMode.code.format("Make It So") == "Make It So")
    }

    @Test func codeModeTrimsTrailingWhitespaceBeforeStrippingPeriod() {
        #expect(DictationMode.code.format("git status.  ") == "git status")
        #expect(DictationMode.code.format("git status  ") == "git status")
    }

    @Test func codeModeLeavesTextWithoutTrailingPeriodAlone() {
        #expect(DictationMode.code.format("git status") == "git status")
    }

    // MARK: - Email mode formatting

    @Test func emailModeAppendsPeriodAfterLetterOrDigit() {
        #expect(DictationMode.email.format("Thanks for the update") == "Thanks for the update.")
        #expect(DictationMode.email.format("We ship on day 3") == "We ship on day 3.")
    }

    @Test func emailModeLeavesSentencePunctuationUntouched() {
        #expect(DictationMode.email.format("Thanks.") == "Thanks.")
        #expect(DictationMode.email.format("Really?") == "Really?")
        #expect(DictationMode.email.format("Wow!") == "Wow!")
        #expect(DictationMode.email.format("Here you go:") == "Here you go:")
        #expect(DictationMode.email.format("Well…") == "Well…")
    }

    @Test func emailModeTrimsTrailingWhitespaceBeforeAppending() {
        #expect(DictationMode.email.format("Thanks for the update  ") == "Thanks for the update.")
    }

    // MARK: - Standard mode formatting

    @Test func standardModeReturnsInputUnchanged() {
        #expect(DictationMode.standard.format("Hello there") == "Hello there")
        #expect(DictationMode.standard.format("Hello there.") == "Hello there.")
        #expect(DictationMode.standard.format("  spaced  ") == "  spaced  ")
    }

    // MARK: - Degenerate input

    @Test func everyModeReturnsEmptyStringForEmptyInput() {
        for mode in [DictationMode.code, .email, .standard] {
            #expect(mode.format("") == "", "expected empty output for \(mode)")
        }
    }

    @Test func everyModeHandlesWhitespaceOnlyInputWithoutAddingPunctuation() {
        for mode in [DictationMode.code, .email, .standard] {
            #expect(!mode.format("   ").contains("."), "expected no punctuation added for \(mode)")
        }
    }
}

File: windows/VoiceKey.Core.Tests/DictationModeTests.cs
namespace VoiceKey.Core.Tests;

public class DictationModeTests
{
    // MARK: - Dictating while VoiceKey itself is frontmost

    [Fact]
    public void ADictationIsInsertedIntoWhateverOtherAppIsFrontmost()
    {
        Assert.True(DictationModes.InsertsAtCursor("Code", "VoiceKey"));
    }

    [Fact]
    public void ADictationTakenWhileVoiceKeyIsFrontmostIsNotInserted()
    {
        Assert.False(DictationModes.InsertsAtCursor("VoiceKey", "VoiceKey"));
    }

    [Fact]
    public void TheFrontmostAppIsMatchedWithoutRegardToCase()
    {
        Assert.False(DictationModes.InsertsAtCursor("voicekey", "VoiceKey"));
    }

    [Fact]
    public void AnUnknownForegroundAppIsAssumedToBeSomewhereElse()
    {
        Assert.True(DictationModes.InsertsAtCursor(null, "VoiceKey"));
        Assert.True(DictationModes.InsertsAtCursor("Code", null));
    }

    // MARK: - Process name → mode mapping

    [Fact]
    public void TerminalsAndEditorsMapToCodeMode()
    {
        string[] codeProcesses =
        [
            "WindowsTerminal.exe",
            "cmd.exe",
            "powershell.exe",
            "pwsh.exe",
            "alacritty.exe",
            "Code.exe",
            "Cursor.exe",
            "devenv.exe",
            "Zed.exe",
            "sublime_text.exe",
            "GitHubDesktop.exe",
        ];

        foreach (var process in codeProcesses)
            Assert.Equal(DictationMode.Code, DictationModes.ForProcess(process));
    }

    [Fact]
    public void JetBrainsIdesMapToCodeMode()
    {
        Assert.Equal(DictationMode.Code, DictationModes.ForProcess("idea64.exe"));
        Assert.Equal(DictationMode.Code, DictationModes.ForProcess("pycharm64.exe"));
        Assert.Equal(DictationMode.Code, DictationModes.ForProcess("rider64.exe"));
    }

    [Fact]
    public void MailClientsMapToEmailMode()
    {
        string[] emailProcesses = ["OUTLOOK.EXE", "olk.exe", "thunderbird.exe", "HxOutlook.exe"];

        foreach (var process in emailProcesses)
            Assert.Equal(DictationMode.Email, DictationModes.ForProcess(process));
    }

    [Fact]
    public void UnknownAppsMapToStandardMode()
    {
        Assert.Equal(DictationMode.Standard, DictationModes.ForProcess("msedge.exe"));
        Assert.Equal(DictationMode.Standard, DictationModes.ForProcess("slack.exe"));
    }

    [Fact]
    public void MissingProcessNameMapsToStandardMode()
    {
        Assert.Equal(DictationMode.Standard, DictationModes.ForProcess(null));
        Assert.Equal(DictationMode.Standard, DictationModes.ForProcess(""));
    }

    [Fact]
    public void ProcessMatchingIgnoresCase()
    {
        Assert.Equal(DictationMode.Code, DictationModes.ForProcess("CODE.EXE"));
        Assert.Equal(DictationMode.Code, DictationModes.ForProcess("code.exe"));
        Assert.Equal(DictationMode.Email, DictationModes.ForProcess("Outlook.exe"));
    }

    [Fact]
    public void ProcessMatchingAcceptsNamesWithAndWithoutTheExeSuffix()
    {
        // Win32 hands back "Code" from the process table but "Code.exe" from a
        // module path; both name the same app.
        Assert.Equal(DictationMode.Code, DictationModes.ForProcess("Code"));
        Assert.Equal(DictationMode.Email, DictationModes.ForProcess("outlook"));
    }

    [Fact]
    public void ProcessMatchingIgnoresAnyLeadingPath()
    {
        Assert.Equal(DictationMode.Code,
            DictationModes.ForProcess(@"C:\Users\me\AppData\Local\Programs\Microsoft VS Code\Code.exe"));
    }

    [Fact]
    public void LookalikeProcessNamesDoNotMatch()
    {
        Assert.Equal(DictationMode.Standard, DictationModes.ForProcess("codecs.exe"));
        Assert.Equal(DictationMode.Standard, DictationModes.ForProcess("outlookhelper.exe"));
    }

    // MARK: - Code mode formatting

    [Fact]
    public void CodeModeStripsSingleTrailingPeriod()
    {
        Assert.Equal("git status", DictationMode.Code.Format("git status."));
    }

    [Fact]
    public void CodeModePreservesEllipsisAndRepeatedDots()
    {
        Assert.Equal("wait...", DictationMode.Code.Format("wait..."));
        Assert.Equal("wait…", DictationMode.Code.Format("wait…"));
    }

    [Fact]
    public void CodeModePreservesOtherTerminalPunctuation()
    {
        Assert.Equal("really?", DictationMode.Code.Format("really?"));
        Assert.Equal("stop!", DictationMode.Code.Format("stop!"));
    }

    [Fact]
    public void CodeModePreservesInteriorPeriodsAndCasing()
    {
        Assert.Equal("node.js is fine", DictationMode.Code.Format("node.js is fine."));
        Assert.Equal("Make It So", DictationMode.Code.Format("Make It So"));
    }

    [Fact]
    public void CodeModeTrimsTrailingWhitespaceBeforeStrippingPeriod()
    {
        Assert.Equal("git status", DictationMode.Code.Format("git status.  "));
        Assert.Equal("git status", DictationMode.Code.Format("git status  "));
    }

    [Fact]
    public void CodeModeLeavesTextWithoutTrailingPeriodAlone()
    {
        Assert.Equal("git status", DictationMode.Code.Format("git status"));
    }

    // MARK: - Email mode formatting

    [Fact]
    public void EmailModeAppendsPeriodAfterLetterOrDigit()
    {
        Assert.Equal("Thanks for the update.", DictationMode.Email.Format("Thanks for the update"));
        Assert.Equal("We ship on day 3.", DictationMode.Email.Format("We ship on day 3"));
    }

    [Fact]
    public void EmailModeLeavesSentencePunctuationUntouched()
    {
        Assert.Equal("Thanks.", DictationMode.Email.Format("Thanks."));
        Assert.Equal("Really?", DictationMode.Email.Format("Really?"));
        Assert.Equal("Wow!", DictationMode.Email.Format("Wow!"));
        Assert.Equal("Here you go:", DictationMode.Email.Format("Here you go:"));
        Assert.Equal("Well…", DictationMode.Email.Format("Well…"));
    }

    [Fact]
    public void EmailModeTrimsTrailingWhitespaceBeforeAppending()
    {
        Assert.Equal("Thanks for the update.", DictationMode.Email.Format("Thanks for the update  "));
    }

    // MARK: - Standard mode formatting

    [Fact]
    public void StandardModeReturnsInputUnchanged()
    {
        Assert.Equal("Hello there", DictationMode.Standard.Format("Hello there"));
        Assert.Equal("Hello there.", DictationMode.Standard.Format("Hello there."));
        Assert.Equal("  spaced  ", DictationMode.Standard.Format("  spaced  "));
    }

    // MARK: - Degenerate input

    [Fact]
    public void EveryModeReturnsEmptyStringForEmptyInput()
    {
        foreach (var mode in AllModes) Assert.Equal("", mode.Format(""));
    }

    [Fact]
    public void EveryModeHandlesWhitespaceOnlyInputWithoutAddingPunctuation()
    {
        foreach (var mode in AllModes) Assert.DoesNotContain(".", mode.Format("   "));
    }

    private static readonly DictationMode[] AllModes =
        [DictationMode.Code, DictationMode.Email, DictationMode.Standard];
}
